src/main/java/com/eerussianguy/blazemap/engine/cache/ChunkMDCache.java

```diff
@@ -65,44 +160,60 @@ public Persisted(RegionMDCache parent) {
             this.parent = parent;
         }
 
+        /**
+         * Read from stream into cache
+         */
         public void read(MinecraftStreams.Input stream) throws IOException {
             int entries = stream.readInt();
 
-            for(int i = 0; i < entries; i++) { // Why can't I just write Key<DataType<>> and have the compiler trust me? ffs.
-                BlazeRegistry.Key<DataType> key = (BlazeRegistry.Key<DataType>) (Object) BlazeMapAPI.MASTER_DATA.findOrCreate(stream.readResourceLocation());
+            // Yes I see the irony that the read() function needs a write lock and the write() function needs a read lock
+            wLock.lock();
+            try {
+                for(int i = 0; i < entries; i++) { // Why can't I just write Key<DataType<>> and have the compiler trust me? ffs.
+                    BlazeRegistry.Key<DataType> key = (BlazeRegistry.Key<DataType>) (Object) BlazeMapAPI.MASTER_DATA.findOrCreate(stream.readResourceLocation());
+
+                    if (key.value() == null) {
+                        BlazeMap.LOGGER.warn("Unrecognised DataType in MD cache. Did you uninstall a Blaze Map extension? Skipping cache read");
+                        throw new IOException("Cannot deserialize unregistered MD DataType");
```

Comment:
We should have a note somewhere that this is a poor solution and needs to be revisited.
But I don't actually require changes to this code in this PR. We just need to get to it later.

Reply:
Have created a ticket: https://blazemap.atlassian.net/browse/BME-188